### PR TITLE
Add parallel chunking support in GPU variant of the numpy reader operator

### PR DESCRIPTION
**Pull request by azrael417**
#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed because of Unet3d: threading intra-sample

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     implemented pread style cufile read
 - Affected modules and functionalities:
    base readers, numpy reader gpu backend
 - Key points relevant for the review:
    handling file offsets when doing threaded reads, more low lying IO routines have to be moved to the loader and/or reader
 - Validation and testing:
    should be transparently covered by existing tests
 - Documentation (including examples):
    same as before

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 1]
```diff
@@ -52,25 +52,43 @@ void NumpyReaderGPU::Prefetch() {
...
   // read the data
   for (size_t data_idx = 0; data_idx < curr_tensor_list.ntensor(); ++data_idx) {
-    thread_pool_.AddWork([&curr_batch, &curr_tensor_list, data_idx](int tid) {
-        curr_batch[data_idx]->read_sample_f(curr_tensor_list.raw_mutable_data(),
-                                            curr_tensor_list.tensor_offset(data_idx) *
-                                            curr_tensor_list.type().size(),
-                                            curr_tensor_list.nbytes());
+    size_t image_bytes = static_cast<size_t>(volume(curr_tensor_list.tensor_shape(data_idx)) \
+                                             * curr_tensor_list.type().size());
+    uint8_t* base_ptr = static_cast<uint8_t*>(curr_tensor_list.raw_mutable_tensor(data_idx));
```
Comment (JanuszL) [inline dali/operators/reader/numpy_reader_gpu_op.cc:68]:
```suggestion
    uint8_t* dst_ptr = static_cast<uint8_t*>(curr_tensor_list.raw_mutable_tensor(data_idx));
```

File: dali/operators/reader/loader/numpy_loader_gpu.cc [hunk 2]
```diff
@@ -105,12 +105,12 @@ void NumpyLoaderGPU::ReadSample(ImageFileWrapperGPU& imfile) {
...
-  imfile.read_sample_f = [this, image_file, &imfile] (void *buffer, Index offset,
-                                                      size_t total_size) {
+  imfile.read_sample_f = [this, image_file, &imfile] (void *buffer, Index file_offset,
+                                                      size_t read_size) {
     // read sample
-    ReadSampleHelper(imfile.file_stream.get(), imfile, buffer, offset, total_size);
+    ReadSampleHelper(imfile.file_stream.get(), imfile, buffer, file_offset, read_size);
     // close the file handle
-    imfile.file_stream->Close();
+    // imfile.file_stream->Close();
```
Comment (JanuszL) [inline dali/operators/reader/loader/numpy_loader_gpu.cc:113]: When the file is closed now?

File: include/dali/core/dynlink_cufile.h [hunk 3]
```diff
@@ -62,7 +62,7 @@ class CufileError : public std::runtime_error {
         return "cufile: invalid pointer memory type";
       case CU_FILE_CUDA_POINTER_RANGE_ERROR:
         return "cufile: pointer range exceeds allocated address range";
-      case CU_FILE_CUDA_CONTECUFILE_OPT_MISMATCH:
+      case CU_FILE_CUDA_CONTEXT_MISMATCH:
```
Comment (JanuszL) [inline include/dali/core/dynlink_cufile.h:65]: We still use old headers in our build. Please revert for now.

File: dali/util/std_cufile.cc [hunk 4]
```diff
@@ -115,6 +115,36 @@ void StdCUFileStream::HandleIOError(int64 ret) const {
...
+  // effective pos:
+  size_t eff_pos = pos_ + file_offset;
+
+  // compute size
+  n_bytes = std::min(n_bytes, length_ - eff_pos);
+
+  // read data: backup n_bytes here and create a read-offset
+  ssize_t n_read = n_bytes;
+  off_t read_off = 0;
+  off_t buffer_off = buffer_offset;
```
Comment (JanuszL) [inline dali/util/std_cufile.cc:129]: Do you need this alias here?

File: dali/operators/reader/loader/numpy_loader_gpu.cc [hunk 5]
```diff
@@ -46,14 +46,14 @@ void NumpyLoaderGPU::RegisterBuffer(void *buffer, size_t total_size) {
...
 
 void NumpyLoaderGPU::ReadSampleHelper(CUFileStream *file, ImageFileWrapperGPU& imfile,
-                                      void *buffer, Index offset, size_t total_size) {
+                                      void *buffer, Index file_offset, size_t read_size) {
   // register the buffer (if needed)
-  RegisterBuffer(buffer, total_size);
+  RegisterBuffer(buffer, read_size);
 
-  Index image_bytes = volume(imfile.shape) * imfile.type_info.size();
+  // Index image_bytes = volume(imfile.shape) * imfile.type_info.size();
```
Comment (JanuszL) [inline dali/operators/reader/loader/numpy_loader_gpu.cc:53]: Could you remove commented out code?

File: dali/operators/reader/loader/numpy_loader_gpu.cc (hunk 2 above)
Comment (azrael417) [reply dali/operators/reader/loader/numpy_loader_gpu.cc:113]: It is closed in:

dali/operators/reader/numpy_reader_gpu_op.cc 

line 89ff

Comment (azrael417) [reply dali/operators/reader/loader/numpy_loader_gpu.cc:113]: I know it is somehow not great to split it that way, so maybe there is some better, callback based method. I leave that to you to improve design, this is more of a functional demo.

File: dali/util/std_cufile.cc (hunk 4 above)
Comment (azrael417) [reply dali/util/std_cufile.cc:129]: you mean buffer_off?  I think that can go in favor of buffer_offset.

Comment (jantonguirao) [reply dali/util/std_cufile.cc:129]: It's incremented in line 138, so I believe it is fine to have it separated.

File: dali/operators/reader/loader/numpy_loader_gpu.cc (hunk 5 above)
Comment (azrael417) [reply dali/operators/reader/loader/numpy_loader_gpu.cc:53]: fixed

File: dali/util/std_cufile.h [hunk 6]
```diff
@@ -33,6 +33,8 @@ class StdCUFileStream : public CUFileStream {
   explicit StdCUFileStream(const std::string& path);
   void Close() override;
   shared_ptr<void> Get(size_t n_bytes) override;
+  size_t ReadGPUThread(uint8_t* gpu_buffer, size_t n_bytes, \
```
Comment (jantonguirao) [inline dali/util/std_cufile.h:36]:
```suggestion
  size_t ReadGPUThread(uint8_t* gpu_buffer, size_t n_bytes,
```
I believe it is not necessary.

File: dali/util/std_cufile.cc [hunk 7]
```diff
@@ -115,6 +115,36 @@ void StdCUFileStream::HandleIOError(int64 ret) const {
   }
 }
 
+size_t StdCUFileStream::ReadGPUThread(uint8_t* gpu_buffer, size_t n_bytes, \
```
Comment (jantonguirao) [inline dali/util/std_cufile.cc:118]: I see that a lot of the code in `ReadGPUThread` is duplicated from `ReadGPU`. Is there a chance we can reuse?

File: dali/util/cufile.h [hunk 8]
```diff
@@ -46,6 +46,8 @@ class DLL_PUBLIC CUFileStream : public FileStream {
    * buffer and the offset where it should put the data.
    */
   virtual size_t ReadGPU(uint8_t* buffer, size_t n_bytes, size_t offset = 0) = 0;
+  virtual size_t ReadGPUThread(uint8_t* buffer, size_t n_bytes, \
```
Comment (jantonguirao) [inline dali/util/cufile.h:49]:
```suggestion
  virtual size_t ReadGPUThread(uint8_t* buffer, size_t n_bytes,
```
nitpick: I think it's not needed. Right?

File: dali/util/std_cufile.cc (hunk 4 above)
Comment (JanuszL) [reply dali/util/std_cufile.cc:129]: But get the argument by value so it is fine to touch it.
Up to you.

File: dali/operators/reader/numpy_reader_gpu_op.cc (hunk 1 above)
Comment (azrael417) [reply dali/operators/reader/numpy_reader_gpu_op.cc]: resolved

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 9]
```diff
@@ -52,25 +52,43 @@ void NumpyReaderGPU::Prefetch() {
...
+
   // read the data
   for (size_t data_idx = 0; data_idx < curr_tensor_list.ntensor(); ++data_idx) {
-    thread_pool_.AddWork([&curr_batch, &curr_tensor_list, data_idx](int tid) {
-        curr_batch[data_idx]->read_sample_f(curr_tensor_list.raw_mutable_data(),
-                                            curr_tensor_list.tensor_offset(data_idx) *
-                                            curr_tensor_list.type().size(),
-                                            curr_tensor_list.nbytes());
+    size_t image_bytes = static_cast<size_t>(volume(curr_tensor_list.tensor_shape(data_idx)) \
+                                             * curr_tensor_list.type().size());
```
Comment (jantonguirao) [inline dali/operators/reader/numpy_reader_gpu_op.cc:67]:
```suggestion
    size_t image_bytes = static_cast<size_t>(curr_tensor_list.tensor_size(data_idx) * curr_tensor_list.type().size());
```
nitpick: just a suggestion.

File: include/dali/core/dynlink_cufile.h (hunk 3 above)
Comment (azrael417) [reply include/dali/core/dynlink_cufile.h:65]: wouldn't it be better to upgrade the header to 0.95? I think its much cleaner

File: dali/util/std_cufile.h (hunk 6 above)
Comment (azrael417) [reply dali/util/std_cufile.h:36]: fixed

File: dali/util/std_cufile.cc (hunk 7 above)
Comment (azrael417) [reply dali/util/std_cufile.cc:118]: read gpu could call read gpu thread in a trivial manner

Comment (azrael417) [reply dali/util/std_cufile.cc]: I fixed it, please review

File: dali/util/std_cufile.cc [hunk 10]
```diff
@@ -115,29 +115,39 @@ void StdCUFileStream::HandleIOError(int64 ret) const {
   }
 }
 
-size_t StdCUFileStream::ReadGPU(uint8_t* gpu_buffer, size_t n_bytes, size_t offset) {
+size_t StdCUFileStream::ReadGPUThread(uint8_t* gpu_buffer, size_t n_bytes,
```
Comment (jantonguirao) [inline dali/util/std_cufile.cc:118]: nitpick: I'd call this `ReadGPUImpl` or similar, since it doesn't contain anything thread specific.

**Review by jantonguirao — APPROVED** [review]

File: include/dali/core/dynlink_cufile.h (hunk 3 above)
Comment (JanuszL) [reply include/dali/core/dynlink_cufile.h:65]: Will do

File: dali/operators/reader/loader/numpy_loader_gpu.cc [hunk 11]
```diff
@@ -46,14 +46,12 @@ void NumpyLoaderGPU::RegisterBuffer(void *buffer, size_t total_size) {
 }
 
 void NumpyLoaderGPU::ReadSampleHelper(CUFileStream *file, ImageFileWrapperGPU& imfile,
-                                      void *buffer, Index offset, size_t total_size) {
+                                      void *buffer, Index file_offset, size_t read_size) {
```
Comment (JanuszL) [inline dali/operators/reader/loader/numpy_loader_gpu.cc:49]: `imfile` is not used anymore.

**Review by JanuszL — APPROVED** [review]

File: dali/util/std_cufile.cc (hunk 10 above)
Comment (azrael417) [reply dali/util/std_cufile.cc]: fixed

File: dali/operators/reader/loader/numpy_loader_gpu.cc (hunk 11 above)
Comment (azrael417) [reply dali/operators/reader/loader/numpy_loader_gpu.cc:49]: removed

**Review by jantonguirao — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**